spacy/tests/serialize/test_serialize_span_groups.py

```diff
@@ -0,0 +1,160 @@
+        (
+            b"\x95\xc4m\x83\xa4name\xa4key1\xa5attrs\x80\xa5spans\x92\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x03\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x04\x00\x00\x00\x07\xc4l\x83\xa4name\xa3too\xa5attrs\x80\xa5spans\x92\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x04\x00\x00\x00\t\x00\x00\x00\x0e\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x05\x00\x00\x00\x0f\x00\x00\x00\x12\xc4l\x83\xa4name\xa3too\xa5attrs\x80\xa5spans\x92\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x04\x00\x00\x00\x07\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x03\xc4C\x83\xa4name\xa4key4\xa5attrs\x80\xa5spans\x91\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x03\xc4C\x83\xa4name\xa4key4\xa5attrs\x80\xa5spans\x91\xc4(\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x03",
+            "How now, brown cow?",
+            {
+                "key1": {"name": "key1", "spans": [(0, 1), (1, 2)]},
+                "too": {"name": "too", "spans": [(1, 2), (0, 1)]},
+                "key4": {"name": "key4", "spans": [(0, 1)]},
+            },
+            True,
+        ),
+    ],
+)
+def test_deserialize_span_groups_compat(
+    en_tokenizer, spans_bytes, doc_text, expected_spangroups, expected_warning
+):
+    """Test backwards-compatibility of `SpanGroups` deserialization.
+    This uses serializations (bytes) from a prior version of spaCy.
+
+    spans_bytes (bytes): Serialized `SpanGroups` object.
+    doc_text (str): Doc text.
+    expected_spangroups (dict):
+        Dict mapping every expected (after deserialization) `SpanGroups` key
+        to a SpanGroup's "args", where a SpanGroup's args are given as a dict:
+          {"name": span_group.name,
+           "spans": [(span0.start, span0.end), ...]}
+    expected_warning (bool): Whether a warning is to be expected from .from_bytes()
+        --i.e. if more than 1 SpanGroup has the same .name within the `SpanGroups`.
+    """
+    doc = en_tokenizer(doc_text)
+
+    if expected_warning:
+        with pytest.warns(UserWarning):
+            doc.spans.from_bytes(spans_bytes)
+    else:
+        doc.spans.from_bytes(spans_bytes)
```

Comment:
Correct me if I'm wrong, but this doesn't catch the behaviour if `expected_warning` is `False` but the deserialization does in fact throw a warning, right?

(I wouldn't hold off on merging because of this though, it's just a very minor thing)

Reply:
Oof. Good catch. I thought I had tested for that when I first wrote that test function... (Maybe my pytest config was turning those warnings into errors? Doubtful; it was probably just PEBKAC. Either way, it shouldn't happen again.)

You're right, and I've written a couple of lines to ensure that warnings aren't ~raised when we don't expect them.

Reply:
(I pushed a commit with those changes to this same branch, but let me know if they should be elsewhere.)

Reply:
We have several places in the test suite that check for no warnings that should probably all be refactored together with a correspond change in the pytest requirements. I've added a link to this discussion in the comments for our internal task so this can be done in the combination with the existing cases. Thanks for your attention to this!